File: packages/types/src/balances.test.ts
import { netBalance } from './balances';

describe('netBalance', () => {
  it('returns zero for no entries', () => {
    expect(netBalance([])).toBe(0n);
  });

  it('is positive when A owes B (a single expense A did not pay for)', () => {
    // Dinner paid by B, A's share is 500 -> A owes B 500.
    expect(netBalance([{ owedByA: true, amount: 500n }])).toBe(500n);
  });

  it('is negative when B owes A', () => {
    expect(netBalance([{ owedByA: false, amount: 200n }])).toBe(-200n);
  });

  it('nets opposing debts from separate expenses (PRD §16 example: owe 500, owed 200 -> net owe 300)', () => {
    const entries = [
      { owedByA: true, amount: 500n },
      { owedByA: false, amount: 200n },
    ];
    expect(netBalance(entries)).toBe(300n);
  });

  it('a settlement entry fully cancels a prior debt (PRD §19/§74)', () => {
    // A owed B 100 from an expense, then A settles 100 with B -- the
    // settlement contributes an entry in the *opposite* direction, which
    // is how SettlementsService constructs it (see its integration spec).
    const entries = [
      { owedByA: true, amount: 100n }, // the original expense debt
      { owedByA: false, amount: 100n }, // the settlement cancelling it
    ];
    expect(netBalance(entries)).toBe(0n);
  });

  it('a partial settlement leaves the remainder outstanding (PRD §75)', () => {
    const entries = [
      { owedByA: false, amount: 500n }, // A paid, B owes A 500
      { owedByA: true, amount: 200n }, // B settles 200 with A -> A owes B 200 entry
    ];
    expect(netBalance(entries)).toBe(-300n); // B still owes A 300
  });

  it('handles many small entries without losing precision', () => {
    const entries = Array.from({ length: 1000 }, (_, i) => ({
      owedByA: i % 2 === 0,
      amount: BigInt(i + 1),
    }));
    const expected = entries.reduce((net, e) => (e.owedByA ? net + e.amount : net - e.amount), 0n);
    expect(netBalance(entries)).toBe(expected);
  });
});

File: apps/api/internal/money/balances_test.go
package money

import (
	"testing"

	"github.com/stretchr/testify/assert"
)

func TestNetBalance(t *testing.T) {
	t.Run("returns zero for no entries", func(t *testing.T) {
		assert.Equal(t, MinorUnits(0), NetBalance(nil))
	})

	t.Run("is positive when A owes B (a single expense A did not pay for)", func(t *testing.T) {
		assert.Equal(t, MinorUnits(500), NetBalance([]LedgerEntry{{OwedByA: true, Amount: 500}}))
	})

	t.Run("is negative when B owes A", func(t *testing.T) {
		assert.Equal(t, MinorUnits(-200), NetBalance([]LedgerEntry{{OwedByA: false, Amount: 200}}))
	})

	t.Run("nets opposing debts from separate expenses (PRD §16 example: owe 500, owed 200 -> net owe 300)", func(t *testing.T) {
		entries := []LedgerEntry{{OwedByA: true, Amount: 500}, {OwedByA: false, Amount: 200}}
		assert.Equal(t, MinorUnits(300), NetBalance(entries))
	})

	t.Run("a settlement entry fully cancels a prior debt (PRD §19/§74)", func(t *testing.T) {
		entries := []LedgerEntry{{OwedByA: true, Amount: 100}, {OwedByA: false, Amount: 100}}
		assert.Equal(t, MinorUnits(0), NetBalance(entries))
	})

	t.Run("a partial settlement leaves the remainder outstanding (PRD §75)", func(t *testing.T) {
		entries := []LedgerEntry{{OwedByA: false, Amount: 500}, {OwedByA: true, Amount: 200}}
		assert.Equal(t, MinorUnits(-300), NetBalance(entries))
	})

	t.Run("handles many small entries without losing precision", func(t *testing.T) {
		entries := make([]LedgerEntry, 1000)
		var expected MinorUnits
		for i := 0; i < 1000; i++ {
			entries[i] = LedgerEntry{OwedByA: i%2 == 0, Amount: MinorUnits(i + 1)}
			if entries[i].OwedByA {
				expected += entries[i].Amount
			} else {
				expected -= entries[i].Amount
			}
		}
		assert.Equal(t, expected, NetBalance(entries))
	})
}
